ugc/serdes_json.hpp

```diff
@@ -0,0 +1,226 @@
+      else if (safe_json && json_is_object(safe_json))
+        json_object_set_new(safe_json.get(), name, m_json.release());
+
+      if (safe_json)
+        m_json = std::move(safe_json);
+    }
+
+    my::JSONPtr m_json = nullptr;
+    Sink & m_sink;
+    HeaderV0 const m_header;
+  };
+
+  template <typename Source>
+  class DeserializerJsonV0
+  {
+  public:
+    DECLARE_EXCEPTION(Exception, RootException);
+
+    DeserializerJsonV0(Source & source, HeaderV0 const & header) : m_source(source), m_header(header)
+    {
+      std::string src;
+      src.resize(source.Size());
+      source.Read(static_cast<void *>(&src[0]), source.Size());
+      m_jsonObject.Attach(src.c_str());
+      m_json = m_jsonObject.get();
+    }
+
+    void operator()(bool & d, char const * name = nullptr)
+    {
+      FromJSONObject(m_json, name, d);
+    }
+
+    void operator()(uint8_t & d, char const * name = nullptr)
+    {
+      FromJSONObject(m_json, name, d);
+    }
+
+    void operator()(uint32_t & d, char const * name = nullptr)
+    {
+      FromJSONObject(m_json, name, d);
+    }
+
+    void operator()(uint64_t & d, char const * name = nullptr)
+    {
+      FromJSONObject(m_json, name, d);
+    }
+
+    void operator()(std::string & s, char const * name = nullptr)
+    {
+      FromJSONObject(m_json, name, s);
+    }
+
+    void operator()(Time & t, char const * name = nullptr)
+    {
+      uint32_t d = 0;
+      FromJSONObject(m_json, name, d);
+      t = FromDaysSinceEpoch(d);
+    }
+
+    void operator()(Sentiment & sentiment, char const * name = nullptr)
+    {
+      bool s = false;
+      FromJSONObject(m_json, name, s);
+      sentiment = s ? Sentiment::Positive : Sentiment::Negative;
+    }
+
+    template <typename T>
+    void operator()(vector<T> & vs, char const * name = nullptr)
+    {
+      json_t * context = SaveContext(name);
+
+      if (!json_is_array(m_json))
+        MYTHROW(my::Json::Exception, ("The field", name, "must contain a json array."));
+
+      vs.resize(json_array_size(m_json));
+      for (size_t index = 0; index < vs.size(); ++index)
+      {
+        json_t * context = SaveContext();
+        m_json = json_array_get(context, index);
+        (*this)(vs[index]);
+        RestoreContext(context);
+      }
+
+      RestoreContext(context);
+    }
+
+    template <typename R>
+    void operator()(R & r, char const * name = nullptr)
+    {
+      json_t * context = SaveContext(name);
+      r.Visit(*this);
+      RestoreContext(context);
+    }
+
+    void VisitRating(float & f, char const * name = nullptr)
+    {
+      double d = 0.0;
+      FromJSONObject(m_json, name, d);
+      f = static_cast<float>(d);
+    }
+
+  private:
+    json_t * SaveContext(char const * name = nullptr)
+    {
+      json_t * context = m_json;
+      if (name)
+        m_json = my::GetJSONObligatoryField(context, name);
+      return context;
+    }
+
+    void RestoreContext(json_t * context)
```

Comment:
As there is always a pair of SaveContext - RestoreContext, why not to:
```c++
template <typename Fn>
void GoToSubtree(Fn && fn)
{
  json_t * json = m_json;
  m_json = my::GetJSONObligatoryField(context, name);
  fn();
  m_json = json;  
}
```
and use it?